src/main/java/net/neoforged/neoforge/transfer/items/templates/ItemStorage.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.items.templates;
+
+import net.minecraft.core.NonNullList;
+import net.minecraft.core.component.DataComponentType;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.component.ItemContainerContents;
+import net.neoforged.neoforge.attachment.AttachmentHolder;
+import net.neoforged.neoforge.attachment.AttachmentType;
+import net.neoforged.neoforge.transfer.ResourceStack;
+import net.neoforged.neoforge.transfer.TransferAction;
+import net.neoforged.neoforge.transfer.context.IItemContext;
+import net.neoforged.neoforge.transfer.handlers.IResourceHandlerModifiable;
+import net.neoforged.neoforge.transfer.items.ItemResource;
+
+public abstract class ItemStorage implements IResourceHandlerModifiable<ItemResource> {
+    protected final int size;
+
+    public ItemStorage(int size) {
+        this.size = size;
+    }
+
+    public abstract ItemContainerContents getContents();
+
+    public abstract int setAndValidate(ItemContainerContents contents, int changedAmount, TransferAction action);
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public ItemResource getResource(int index) {
+        return getContents().getImmutableStackInSlot(index).resource();
+    }
+
+    @Override
+    public int getAmount(int index) {
+        return getContents().getImmutableStackInSlot(index).amount();
+    }
+
+    @Override
+    public int getCapacity(int index, ItemResource resource) {
+        return resource.getMaxStackSize();
+    }
+
+    @Override
+    public int getCapacity(int index) {
+        return 99;
```

Comment:
Might be best to use the reference value for long term `Item.ABSOLUTE_MAX_STACK_SIZE`